[feature] Add support for "independent" EBIC on SPARC

pieleric opened the pull request:
Add the support for changing the settings of an EBIC detector and playing it live when it's an "independent" detector (ie, it's not directly synchronized with the e-beam scanner component). 
Also support SPARC synchronized acquisition.

To validate this code, introduce a simple simulator. To simulate the hardware, it pretends to not support continuous acquisition.

pieleric committed:
[feature] new simsem.IndependentDetector

Simulate a detector which has resolution and dwellTime info.
Used to simulate the future separate EBIC detector controller.

pieleric committed:
[config] add SPARC simulator with EBIC and CL PMT

pieleric committed:
[feature] extend SPARC GUI to support "independent" EBIC detector

Mostly just a matter to detecting the detector is special and requesting
the corresponding stream type.

pieleric committed:
[feature] SPARC acq: new stream to acquire with an "independent" EBIC

Introduce a new type of "setting" stream, which shows the live view of
the indenpendent EBIC while also starting the e-beam scanning using a second
dataflow, and copy the metadata.

Extend the "standard" synchronized e-beam acquisition to support
independent EBIC. Mostly a matter of copying the .resolution and
.dwellTime beforehand, and copying the metadata of the e-beam data onto
the EBIC data afterwards.

pieleric commented on install/linux/usr/share/odemis/sim/sparc2-independent-ebic-sim.odm.yaml:
```diff
@@ -0,0 +1,357 @@
+}
+
+# Provide the mirror (aka port selector) of the spectrograph as a separate component
+"Spec CCD Flipper": {
```
The Odemis optical path manager doesn't care. However, in reality, typically the spectrograph always has a flipper, and it's important to place it to the right position, because the flipper position is "random" at start-up.

I'd rather keep the simulator as similar as the real hardware, so keep this component.

pieleric commented on src/odemis/acq/stream/_helper.py:
```diff
@@ -1448,9 +1450,136 @@ def __init__(self, name, detector, dataflow, emitter, **kwargs):
+    """
+    def __init__(self, name, detector, dataflow, emitter, emt_dataflow, **kwargs):
+        """
+        emtvas: don't put resolution or scale
```
It's part of the kwargs. I couldn't find an official convention to indicate it with the docstring sphynx format, but I've now clearly indicated it in the message.

pieleric commented on src/odemis/acq/stream/_helper.py:
```diff
@@ -1448,9 +1450,136 @@ def __init__(self, name, detector, dataflow, emitter, **kwargs):
+        # That's because we either update a data which already has the latest metadata, in which
+        # case this doesn't change the metadata, or we update the old data, which will be immediately replaced
+        # by the new one, in which case it's not a big deal.
+        # To be extra careful, we could check the .shape (should be the same) and the MD_ACQ_DATE
```
It's not. It's an idea. I've now put this under a TODO tag.

pieleric commented on src/odemis/acq/stream/_helper.py:
```diff
@@ -1448,9 +1450,136 @@ def __init__(self, name, detector, dataflow, emitter, **kwargs):
         super().__init__(name, detector, dataflow, emitter, **kwargs)
 
+
+SCANNER_POS_MD = {model.MD_POS, model.MD_POS_COR, model.MD_PIXEL_SIZE, model.MD_PIXEL_SIZE_COR,
```
Well, if it's a constant which is used by a single class, I prefer to put it just above (before) that class. This makes it clearer where the constant is being used. See `OVRL_MAX_DIFF` line ~1586.

pieleric commented on src/odemis/driver/simsem.py:
```diff
@@ -543,16 +543,137 @@ def _acquire_thread(self, callback):
+
+    def _wait_acquisition_stopped(self):
+        """
+        Waits until the acquisition thread is fully finished _iff_ it was requested
```
"iff" is an abbreviation for "if and only if". I'll write it fully as it's not super famous.

pieleric commented on src/odemis/driver/simsem.py:
```diff
@@ -543,16 +543,137 @@ def _acquire_thread(self, callback):
+        """
+        # "if" is to not wait if it's already finished
+        if self._acquisition_must_stop.is_set():
+            logging.debug("Waiting for thread to stop.")
+            self._acquisition_thread.join(10)  # 10s timeout for safety
+            if self._acquisition_thread.is_alive():
+                logging.exception("Failed to stop the acquisition thread")
```
It's a code that's very unlikely code to be triggered. I just want to keep it as simple as possible. So I'll just hard-code the "10s" in the logging message.

pieleric commented on src/odemis/driver/simsem.py:
```diff
@@ -543,16 +543,137 @@ def _acquire_thread(self, callback):
+
+        logging.debug("Simulating an image")
+        res = self.resolution.value
+        max_res = self.fake_img.shape[::-1]  # Y, X
+
+        # Take the center of the image, with the specified resolution
+        center = (max_res[1] / 2, max_res[0] / 2)
```
I've finally changed it to use `self.resolution.range[1]`, which makes it less awkward wrt the X/Y order. (the previous code was actually wrong!)

pieleric commented on src/odemis/driver/test/simsem_test.py:
```diff
@@ -518,5 +518,51 @@ def test_shift(self):
+        self.assertEqual(self.det.dwellTime.value, 1e-6)
+
+        self.det.dwellTime.value = 1.16e-6
+        self.assertAlmostEqual(self.det.dwellTime.value, 1.1e-6, delta=0.1e-9)
```
The dwellTime VA is a FloatContinuous, so it'll complain if you pass a value higher than the maximum.
However, I've added a check for the smallest value, to make sure it stays within the range.

pieleric commented on src/odemis/driver/test/simsem_test.py:
```diff
@@ -518,5 +518,51 @@ def test_shift(self):
+            expected_duration = self.compute_expected_duration()
+
+            start = time.time()
+            im = self.det.data.get()
```
`img` is clear enough?

pieleric commented on src/odemis/driver/simsem.py:
```diff
@@ -338,7 +338,7 @@ def __init__(self, name, role, parent, **kwargs):
         """
         # It will set up ._shape and .parent
         model.Detector.__init__(self, name, role, parent=parent, **kwargs)
-        self.data = SEMDataFlow(self, parent)
```
I've changed the `SEMDataFlow` class to not require a second parameter (it actually didn't even use it). This is the only place it was used so far.

pieleric commented:
@sneepstefan , I've fixed the code according to your comments. I've also extended the simulator to have an option to not support continuous acquisition, like the MightyEBIC, and also to add some noise to the image mostly to see better when an new image is generated. 
I've also extend the stream to automatically restart the acquisition, as the hardware doesn't support continuous acquisition by itself. 
Finally, I've added test cases for the new stream.

tmoerkerken commented on src/odemis/acq/stream/_helper.py:
```diff
@@ -1448,9 +1450,161 @@ def __init__(self, name, detector, dataflow, emitter, **kwargs):
         """
         if "acq_type" not in kwargs:
-            kwargs["acq_type"] = model.MD_AT_EM
+            kwargs["acq_type"] = model.MD_AT_EBIC
```
It is a bit unclear to me why this is the fallback now, does it matter? Do we rely on it?

pieleric replied:
We didn't have an office "EBIC" acquisition type until now. So we had been lazy and claim that EBIC is the same as SEM. This PR add the EBIC acquisition type, and here, we retro-actively use it on the (old) standard EBIC stream. Note that it's not very used in Odemis. Mostly for the icon in the merge slider.

tmoerkerken commented on src/odemis/acq/stream/_helper.py:
```diff
@@ -1448,9 +1450,161 @@ def __init__(self, name, detector, dataflow, emitter, **kwargs):
+    """
+    def __init__(self, name, detector, dataflow, emitter, emt_dataflow, **kwargs):
+        """
+        :param emt_dataflow: a DataFlow that can be used to start/stop the emitter
```
The docstring parameters are incomplete (and there is lack of typing in the constructor). It's a bit symptomatic for this class.

pieleric replied:
Indeed. The first 4 parameters are standard parameters for all the Stream classes. They are defined on the base class `Stream`. Repeating them on every of the ~20 classes would be very redundant. I'll add 1 line to mention it though.

tmoerkerken commented on src/odemis/acq/stream/_sync.py:
```diff
@@ -2322,8 +2322,22 @@ def _adjustHardwareSettings(self):
 
+        # If a detector is "independent", then connect its dwell time
+        # Note: resolution is updated later, just before the acquisition block
+        # TODO: should it be done automatically by the "SettingsStream" in .prepare() or .linkHwVA()?
```
When are we going to do all the TODO's? :smile:

pieleric replied:
"One day". Well, for this TODO, I did add a task to the backlog, so this might really happen eventually!
https://delmic.atlassian.net/browse/CSDM-1131

tmoerkerken commented on src/odemis/acq/stream/_sync.py:
```diff
@@ -2420,6 +2434,17 @@ def _runAcquisition(self, future):
+                    det = s._detector
+                    if model.hasVA(det, "resolution"):
+                        det.resolution.value = (n_x, n_y)
+                        if det.resolution.value != (n_x, n_y):
```
This reads quite funnily. I suppose the setter can prevent the actual setting of the value, but it is not really trivial from just reading the code. Maybe comment why you are anticipating a failure of setting the value.

tmoerkerken commented on src/odemis/acq/test/stream_test.py:
```diff
@@ -3914,6 +3915,164 @@ def test_ar_acq_integrated_images_leech(self):
+
+
+
+
```
This is a lot of whitespace!

tmoerkerken commented on src/odemis/driver/simsem.py:
```diff
@@ -543,16 +548,161 @@ def _acquire_thread(self, callback):
+
+        # Take the center of the image, with the specified resolution
+        center = (max_res[0] / 2, max_res[1] / 2)
+        ltrb = [int(center[0] - (res[0] / 2)),
```
This kind of manual code craves for an AABB (axis aligned bounding box) module!
Such a module would have a function like `def expand_rect(rect, margin):` from the `util` module, but those functions would be a lot more generalized an reusable.

tmoerkerken reviewed (changes requested):
Saw GUI and tests in action and all looks good. I have some minor aesthetic comments still.

pieleric committed:
[feat] add EBIC icon to merge slider

Now that there is an official EBIC acquisition type, we also need an
icon in the GUI for the merge slider.
